### PR TITLE
Close message channel issue

### PR DESCRIPTION
**Pull request by omercelikceng**
Resolves https://github.com/spring-cloud/spring-cloud-stream/issues/2869

### CODE REVIEW

**Comment by omercelikceng** [comment]
Alright, @olegz. I’ll explain in detail, but I'm currently on holiday. I'll provide a detailed explanation on Sunday."

**Comment by omercelikceng** [comment]
> @omercelikceng The issue #2869 marked as closed. Can you please explain what you are adding with this PR?

Hi @olegz .  I'd like to explain using the tests I've written. Assume we want to produce data by providing a binder to a dynamic destination. 

image

In this example, when data is sent, it is put into the **channelCache** map within the **StreamBridge** class under the **key** "**kafka1:dynamicTopic**". The relevant code is shown in the image below.

bindProducer

Additionally, the **this.bindingService.bindProducer(..)** method is also invoked here. Let’s examine this method. When this method is called, the data is put into the **producerBindings** map within the **BindingService** class under the **key "dynamicTopic".**  The relevant code is shown in the image below.

bind2


Moreover, all classes derived from **AbstractBindableProxyFactory** populate the **outputHolders map,** where the **key** is solely the **output name.** These classes also have the capability to invoke the **unbindProducer** method.

Subsequently, when the application is shutting down, the **destroy** method of the **StreamBridge** class is executed. There, all relevant values in the **channelCache** map are unbound (via channelCache.keySet().forEach(bindingService::unbindProducers);). However, since the keys in channelCache differ from those in producerBindings, the unbind operation is not performed.

In other words, the key values in these maps are not identical. To resolve this issue during the unbind process, I made several enhancements to correct the inconsistencies. 

**If you run the tests I have written with the previous code, you will see that the error occurs.**
